lib/psych/visitors/yaml_tree.rb

```diff
@@ -310,19 +310,20 @@ def visit_String o
           style = Nodes::Scalar::LITERAL
           plain = false
           quote = false
-        elsif o =~ /\n/
-          style = Nodes::Scalar::LITERAL
         elsif o == '<<'
           style = Nodes::Scalar::SINGLE_QUOTED
           tag   = 'tag:yaml.org,2002:str'
           plain = false
           quote = false
+        elsif o =~ /\n[^\Z]/  # match \n except blank line at the end of string
+          style = Nodes::Scalar::LITERAL
```

Comment:
Can we keep this above the `'<<'` test?  Any string that contains a newline will surely fail the '<<' test, so it would be better if we can keep it early.